Modules/rules.py

```diff
@@ -0,0 +1,132 @@
+            only the word will be matched against.
+        pass_match (bool):
+            If this is True, the match object will be passed as an argument toward the command
+            function.
+        prefixless (bool):
+            If this is True, the rule can match whether or not the prefix is present. Otherwise, it
+            acts similarly to a command, only being considered if the message starts with our
+            prefix. The prefix itself should not be part of the regex for non-prefixless rules.
+        after (Rule):
+            A rule which this one must be considered after, for some rudimentary prioritization.
+
+    Please note that *regex* can match anywhere within the string, it need not match the entire
+    string. If you wish to change this behaviour, use '^' and '$' in your regex.
+
+    Ordering is actually ensured by requiring access to *after*, by making sure that it has been
+    registered before this one.
+
+    Raises:
+        DuplicateRuleException: If the same rule has already been registered.
+        RuleNotPresentException: If the rule *after* hasn't yet been registered.
+    """
+    def decorator(coro: Callable):
+        if case_sensitive:
+            pattern = re.compile(regex)
+        else:
+            pattern = re.compile(regex, re.IGNORECASE)
+
+        tuple_ = Rule(pattern, coro, full_message, pass_match)
+
+        target = _prefixless_rules if prefixless else _rules
+        if tuple_ in target:
+            raise DuplicateRuleException(tuple_)
+
+        if after is None:
+            target.append(tuple_)
+        else:
+            try:
+                target.insert(target.index(after) + 1, tuple_)
+            except ValueError:
+                raise RuleNotPresentException(after)
+
+        log.info(f"New rule matching '{regex}' case-{'' if case_sensitive else 'in'}sensitively was"
+                 f" created.")
+        return tuple_
+    return decorator
+
+
+def get_rule(words: List[str], words_eol: List[str],
+             prefixless: bool) -> Tuple[Optional[Callable], tuple]:
+    """
+    Attempt to find a rule in the given dict of patterns and rules.
+
+    Args:
+        words: Words of the message.
+        words_eol: Words of the message, but each including everything to the end of it.
+        prefixless: Whether or not we're looking for prefixless rules.
+
+    Returns:
+        (Callable or None, tuple):
+            2-tuple of the command function and the extra args that it should
+            be called with.
+    """
+    for (pattern, fun, full_message, pass_match) in (_prefixless_rules if prefixless else _rules):
+        if full_message:
+            match = pattern.match(words_eol[0])
+        else:
+            match = pattern.match(words[0])
+
+        if match is not None:
+            if pass_match:
+                return fun, (match,)
+            else:
+                return fun, ()
+    else:
+        return None, ()
```

Comment:
This expression is always truthy, which [causes issues in other modules when a truthy check is performed.](https://github.com/FuelRats/pipsqueak3/pull/54/files#diff-f3ecc38ebed1a7b33077a674020998efR84)
```py
In [1]: def foo():
   ...:     return None, ()
   ...:
   ...:

In [2]: retn = foo()

In [3]: retn
Out[3]: (None, ())

In [4]: bool(retn)
Out[4]: True

In [5]: bool(foo())
Out[5]: True

In [6]: if foo():
   ...:     print("truthy!")
   ...:
truthy!
```

My suggestion here is to simply return `None` and let the caller check the return with an explicit None check

Reply:
I disagree that this made it less clear. The meaning of the return value is: No rule was found and no arguments should be appended to the call. [Returning None in place of the tuple will make this less clear](http://thefinestartist.com/effective-java/43) in my opinion, and returning just a `None` will obfuscate client code:
```python
result = get_rule(...)
if result is not None:
  fun, extra_args = result
```
instead of 
```python
fun, extra_args = result
```
to achieve the same result.